docs/ref-manual/cvl/invariants.md

```diff
@@ -28,29 +35,269 @@ preserved_block ::= "preserved"
+
+```solidity
+contract Timestamp {
+    uint256 public immutable timestamp;
+
+    constructor() {
+        timestamp = block.timestamp;
+    }
+}
+```
+
+The following invariant will be successfully verified, although it is clearly
+false:
+
+```cvl
+invariant time_is_now(env e)
+    timestamp() == e.block.timestamp;
+```
+
+The verification is successful because the action that falsifies the invariant
+is the passage of time, rather than the invocation of a contract method.
+
+Similarly, an invariant that depends on an external contract can become false
+by calling a method on the external contract.  For example:
+
+```solidity
+contract SupplyTracker {
+    address token;
+    uint256 public supply;
+
+    constructor(address _token) {
+        token  = _token;
+        supply = token.totalSupply();
+    }
+}
+```
+
+As above, an invariant stating that `supply() == token.totalSupply()` would be
+verified, but a method on `token` might change the total supply without updating
+the `SupplyTracker` contract.  Since the Prover only checks the main contract's
+methods for preservation, it will not report that the invariant can be
+falsified.
+
+For this reason, invariants that depend on the environment or on the state of
+external contracts are a potential source of {term}`unsound`ness, and should be
+used with care.
+
+
 Filters
 -------
 
-```{todo}
-This feature is currently undocumented.
+For performance reasons, you may want to avoid checking that an invariant is
+preserved by a particular method or set of methods.  Invariant filters provide
+a method for skipping verification on a method-by-method basis.
+
+```{caution}
+Filtering out methods while checking invariants is {term}`unsound`.
 ```
 
+To filter out methods from an invariant, add a `filtered` block after the
+expression defining the invariant.  The body of the `filtered` block must
+contain a single filter of the form `var -> expr`, where `var` is a variable
+name, and `expr` is a boolean expression that may depend on `var`.
+
+Before verifying that a method `m` preserves an invariant, the `expr` is
+evaluated with `var` bound to a `method` object.  This allows `expr` to refer
+to the fields of `var`, such as `var.selector` and `var.isView`.  See
+{ref}`method-type` for a list of the fields available on `method` objects.
+
+If the expression evaluates to `false` with `var` replaced by a given method,
+the Prover will not check that the method preserves the invariant.  For example,
+the following invariant will not be checked on the `deposit(uint)`
+method:
+
+```cvl
+invariant balance_is_0(address a)
+    balanceOf(a) == 0
+    filtered {
+        f -> f.selector != deposit(uint).selector
+    }
+```
+
+In this example, when the variable `f` is bound to `deposit(uint)`, the
+expression `f.selector != deposit(uint).selector` evaluates to `false`, so the
+method will be skipped.
+
+```{note}
+If there is a {ref}`preserved block <preserved>` for a method it will be
```

Comment:
maybe say that if one complements a filtered expression with preserved blocks for the filtered methods, this is one way to handle the unsoundness originally introduced by the filtered expression.

Reply:
also wondering whether the discussion of preserved blocks shouldn't precede the filtered expressions - but it's fine either way for me

Reply:
Fixed